src/librustc_mir/interpret/eval_context.rs

```diff
@@ -23,14 +23,22 @@ use super::{
     Immediate, MPlaceTy, Machine, MemPlace, Memory, OpTy, Operand, Place, PlaceTy,
     ScalarMaybeUndef, StackPopInfo,
 };
+use rustc::infer::canonical::OriginalQueryValues;
+use rustc::infer::InferCtxt;
+use rustc::traits::ObligationCause;
 
-pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
+pub struct InterpCx<'infcx, 'mir, 'tcx, M: Machine<'mir, 'tcx>> {
     /// Stores the `Machine` instance.
     pub machine: M,
 
     /// The results of the type checker, from rustc.
     pub tcx: TyCtxtAt<'tcx>,
 
+    /// The inference context for inference variables that are within this `InterpCx`. Inference
+    /// variables may be within either the `param_env` or within `substs` of frames. Many operations
+    /// outside of mir which takes values that may contain variable require this context.
```

Comment:
```suggestion
    /// outside of mir which take values that may contain variable require this context.
```
Also, what is "outside of mir"?

And -- does this basically mean we have to use this (one way or another) whenever we use `param_env` or a frame `subst`?